contracts/src/bridge/SequencerInbox.sol

```diff
@@ -172,38 +184,101 @@ contract SequencerInbox is DelegateCallAware, GasRefundEnabled, ISequencerInbox
         );
     }
 
-    function addSequencerL2Batch(
+    function addSequencerL2BatchFromOrigin(
         uint256 sequenceNumber,
         bytes calldata data,
         uint256 afterDelayedMessagesRead,
-        IGasRefunder gasRefunder
+        IGasRefunder gasRefunder,
+        uint256 prevMessageCount,
+        uint256 newMessageCount
     ) external refundsGas(gasRefunder) {
-        if (!isBatchPoster[msg.sender] && msg.sender != address(rollup)) revert NotBatchPoster();
-
+        // solhint-disable-next-line avoid-tx-origin
+        if (msg.sender != tx.origin) revert NotOrigin();
+        if (!isBatchPoster[msg.sender]) revert NotBatchPoster();
         (bytes32 dataHash, TimeBounds memory timeBounds) = formDataHash(
             data,
             afterDelayedMessagesRead
         );
-        // we set the calldata length posted to 0 here since the caller isn't the origin
-        // of the tx, so they might have not paid tx input cost for the calldata
+        // Reformat the stack to prevent "Stack too deep"
```

Comment:
we could update the compiler and compile with `via ir` enabled to avoid stack issues (but we can prob do this in a follow up PR instead)